Rewrite torch.broadcast_shapes to be unbacked SymInt friendly

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95217

This is similar to what I did in https://github.com/pytorch/pytorch/pull/94790/files#diff-39e82af71afdadbc56a4ecf552ec668ddcda794f8ea3ec41edaef23456fd56e9
but I have to do this everywhere we have a broadcasting implementation.

If you want me to spend some BE time deduping these, please holler.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Rewrite torch.broadcast_shapes to be unbacked SymInt friendly

This is similar to what I did in https://github.com/pytorch/pytorch/pull/94790/files#diff-39e82af71afdadbc56a4ecf552ec668ddcda794f8ea3ec41edaef23456fd56e9
but I have to do this everywhere we have a broadcasting implementation.

If you want me to spend some BE time deduping these, please holler.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

albanD reviewed (approved):
That is quite a bit more intrusive than I would like it to be...
Also it has been quite a while without tests for these?

But the change is ok.

ezyang commented:
The E2E is in https://github.com/pytorch/pytorch/pull/95218 but I can make a unit test for this. Actually, I can't even land this PR yet because the PR it was based on got reverted. 

> That is quite a bit more intrusive than I would like it to be...

Let's talk about it. There seem to be two problems:

1. I have to keep doing this change in lots of places. I could marginally make this better by trying to dedupe broadcast implementations.
2. The change itself is quite invasive. I'm not sure of a better way to write out this change. I could factor out the "broadcast two single sizes" into its own helper function, but it'd still have roughly the same shape internally.

WDYT?

ezyang committed:
Update on "Rewrite torch.broadcast_shapes to be unbacked SymInt friendly"


This is similar to what I did in https://github.com/pytorch/pytorch/pull/94790/files#diff-39e82af71afdadbc56a4ecf552ec668ddcda794f8ea3ec41edaef23456fd56e9
but I have to do this everywhere we have a broadcasting implementation.

If you want me to spend some BE time deduping these, please holler.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]
